### PR TITLE
Uninhabited != NoReturn

### PR DESCRIPTION
**Pull request by CodaFi**
#### What's in this pull request?

Fix a case where we would fail to diagnose a missing return if an enum decl as the result of a function had no cases.

@slavapestov can you review?

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by slavapestov** [comment]
Why do we want this?

**Comment by CodaFi** [comment]
There's a clear difference between `Never` the type and any user-defined enumeration, regardless of whether it has cases or not.

**Comment by jrose-apple** [comment]
There's not intended to be. We need to diagnose "falling off the end" in either case.

**Comment by slavapestov** [comment]
Right. To clarify, we want any uninhabited type to behave like Never, and mark a function as not returning.

Right now the isNever check is inaccurate because it doesn't realize that a strict with an uninhabited field is also uninhabited, but that didn't seem worth fixing.

**Comment by slavapestov** [comment]
@jrose-apple By definition we cannot fall off the end if the return type is Never, because the return is not reachable.

**Comment by CodaFi** [comment]
When dealing with user-defined types it's more likely that they meant to write up the cases for the enum later and would expect a diagnostic about the function not returning than they decided to exploit `ex falso`.  There is only one bottom type, and that is `Never`.  If the user wishes to explore alternate expressions of that fact, they can always reach for the classic

```swift
enum X {}

func explode() -> X {
  return explode()
}
```

**Comment by CodaFi** [comment]
@swift-ci please smoke test Linux platform.

**Comment by slavapestov** [comment]
If defer to @jckarter regarding this. It was his original suggestion to treat all empty enums as a 'Never' type.

**Comment by slavapestov** [comment]
Actually, the fact that this doesn't produce a diagnostic is the real bug:

```
enum E {}

func f() -> E {}
```

It happens for both 'E' and 'Never'. Mind if I take this?

File: test/SILOptimizer/return.swift [hunk 1]
```diff
@@ -9,6 +9,11 @@ func singleBlock2() -> Int {
   y += 1
 } // expected-error {{missing return in a function expected to return 'Int'}}
 
+enum NoCasesButNotNever {}
+
+func diagnoseNoCaseEnumMissingReturn() -> NoCasesButNotNever {
```
Comment (slavapestov) [inline test/SILOptimizer/return.swift]: Note that this is meant to be an error even with a 'Never' return. a Never-returning function should end with a call to some other Never-returning function, or the return should be unreachable.

**Comment by CodaFi** [comment]
That it happens for `E` is what this fixes.  For `Never` there shouldn't be a diagnostic, right?

**Comment by slavapestov** [comment]
@CodaFi That's the thing, we should get the diagnostic for Never too.

Remember, this was invalid before:

```
@noreturn func f() {}
```

A @noreturn function had to either end with a call to some other @noreturn function, or end with an infinite loop, essentially. You weren't allowed to fall through the end.

It looks like something is wrong in the SIL unreachable diagnostics pass wrt the Never type.

**Comment by CodaFi** [comment]
Sounds good.

**Pull request closed without merging**